src/vm/object.cpp

```diff
@@ -2051,81 +2051,117 @@ StackTraceElement & StackTraceArray::operator[](size_t index)
 }
 
 #if !defined(DACCESS_COMPILE)
-// Define the lock used to access stacktrace from an exception object
-SpinLock g_StackTraceArrayLock;
 
 void ExceptionObject::SetStackTrace(StackTraceArray const & stackTrace, PTRARRAYREF dynamicMethodArray)
 {        
     CONTRACTL
     {
-        GC_NOTRIGGER;
-        NOTHROW;
+        GC_TRIGGERS;
+        THROWS;
         MODE_COOPERATIVE;
     }
     CONTRACTL_END;
 
-    Thread *m_pThread = GetThread();
-    SpinLock::AcquireLock(&g_StackTraceArrayLock, SPINLOCK_THREAD_PARAM_ONLY_IN_SOME_BUILDS);
+    struct _gc
+    {
+        ExceptionObject exception;
+        I1ARRAYREF stackTraceArray;
+        PTRARRAYREF dynamicMethodArray;
+    } gc;
+    ZeroMemory(&gc, sizeof(gc));
+    GCPROTECT_BEGIN(gc);
 
-    SetObjectReference((OBJECTREF*)&_stackTrace, (OBJECTREF)stackTrace.Get());
-    SetObjectReference((OBJECTREF*)&_dynamicMethods, (OBJECTREF)dynamicMethodArray);
+    gc.exception = this;
+    gc.stackTraceArray = stackTrace.Get();
+    gc.dynamicMethodArray = dynamicMethodArray;
 
-    SpinLock::ReleaseLock(&g_StackTraceArrayLock, SPINLOCK_THREAD_PARAM_ONLY_IN_SOME_BUILDS);
+    _dispatchStateLock->EnterObjMonitor();
 
+    SetObjectReference((OBJECTREF*)&_stackTrace, (OBJECTREF)gc.stackTraceArray);
```

Comment:
This cannot use `this` anymore. It needs to use the protected reference, something like `gc.exception->_stackTrace`.